### PR TITLE
*: initial implementation of grpc-proxy

### PR DESCRIPTION
**Pull request by xiang90**
@mqliang @heyitsanthony

### CODE REVIEW

File: proxy/grpcproxy/cache/store.go [hunk 1]
```diff
@@ -76,15 +78,15 @@ func (c *cache) Get(req *pb.RangeRequest) (*pb.RangeResponse, error) {
 	c.mu.Lock()
 	defer c.mu.Unlock()
 
-	if req.Revision > c.compactedRev {
+	if c.compactedRev != -1 && req.Revision <= c.compactedRev {
```
Comment (heyitsanthony) [inline proxy/grpcproxy/cache/store.go]: `< compactedRev`? -- `compact(rev)` will keep `rev`

File: etcdmain/grpc_proxy.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+func newGRPCProxyStartCommand() *cobra.Command {
+	cmd := cobra.Command{
+		Use:   "start",
+		Short: "start the grpc proxy",
+		Run:   startGRPCProxy,
+	}
+
+	cmd.Flags().StringVar(&grpcProxyListenAddr, "listen-addr", "127.0.0.1:23790", "listen address")
+	cmd.Flags().StringSliceVar(&grpcProxyEndpoints, "endpoints", []string{"127.0.0.1:2379"}, "comma separated etcd cluster endpoints")
+
```
Comment (heyitsanthony) [inline etcdmain/grpc_proxy.go]: tls?

File: etcdmain/grpc_proxy.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	client, err := clientv3.New(cfg)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
+	}
+
+	kvp := grpcproxy.NewKvProxy(client)
+
+	var opts []grpc.ServerOption
+	server := grpc.NewServer(opts...)
```
Comment (gyuho) [inline etcdmain/grpc_proxy.go]: This just can be `grpc.NewServer()` unless we want to have `opts` for later?

File: etcdmain/grpc_proxy.go (hunk 2 above)
Comment (xiang90) [inline etcdmain/grpc_proxy.go]: do you mean tls support for the proxy? leave it as a todo?

File: proxy/grpcproxy/cache/store.go (hunk 1 above)
Comment (xiang90) [inline proxy/grpcproxy/cache/store.go]: oh. right. will fix.

File: etcdmain/grpc_proxy.go (hunk 2 above)
Comment (heyitsanthony) [inline etcdmain/grpc_proxy.go]: might as well have it now, I don't think it's much code?

Comment (heyitsanthony) [inline etcdmain/grpc_proxy.go]: (at least TLS to talk to etcd, not sure about the listener...)

**Comment by xiang90** [comment]
@gyuho @heyitsanthony PTAL

**Comment by gyuho** [comment]
lgtm. defer to @heyitsanthony 

Thanks!

**Comment by heyitsanthony** [comment]
lgtm

**Pull request merged**